### PR TITLE
improve mapValues types

### PR DESCRIPTION
**Pull request by dartess**
Closes https://github.com/remeda/remeda/issues/354

### CODE REVIEW

File: src/_types.ts [hunk 1]
```diff
@@ -21,3 +21,10 @@ export type NonEmptyArray<T> = [T, ...Array<T>];
  * @see This was inspired by the type-definition of Promise.all (https://github.com/microsoft/TypeScript/blob/1df5717b120cddd325deab8b0f2b2c3eecaf2b01/src/lib/es2015.promise.d.ts#L21)
  */
 export type IterableContainer<T = unknown> = ReadonlyArray<T> | [];
+
+// Inspired and largely copied from `sindresorhus/ts-extras`:
+// @see https://github.com/sindresorhus/ts-extras/blob/44f57392c5f027268330771996c4fdf9260b22d6/source/object-keys.ts
+export type ObjectKeys<T extends Record<PropertyKey, unknown>> = `${Exclude<
```
Comment (eranhirsch) [inline src/_types.ts:27]: Any reason to switch from `object` to `Record<PropertyKey, unknown>`. If the typing works for the objects we want, I don't mind us being broader and allowing unusual usages to be applicable too (and if they are buggy, to get an issue report).

https://www.typescriptlang.org/play?#code/C4TwDgpgBA8gRgKwgY2AaQiAzjANgEwB4AVKCAD2AgDt8soB7RFYAPigF4oADAEgG8AouWS4ArvgiEA1pgYAzKMQA0ULCAC2cBrlYBfbgG4AUKEixmqDNgByEAO4kylGnSgAlFAwBORAAreDJDeoNaqYtTS1Az21KzsXHxCIuKSMnKKKmqa2roGJsam4NAAwgCGuLhlcLjQXPzGUFAAFACUAFxQAEQAFhCVDF0mTWWdXfY+BEPGeoXIDNRYwFDkneWV1bWcFkioAHRlWFgAlgDm1M1tnOy9-biDqvxQo90T3lPP9POLwHqtBd8llBvBAsGJcMsuOQ2iYAPSwppNAB6AH5jICdBA9vdTs1yAd-sZ4YiSaTUYV0QsgSBtvBdsA9rJsHj-lBiRx2ABtADkZW5AF1CmZoNYcPJ1lUahA8PhaZZ0JgcARCBLNhBWHCEci0ULilBRTBxRVJbU7PY5fTRWaVca1RrCkA

```suggestion
export type ObjectKeys<T extends object> = `${Exclude<
```

**Review by eranhirsch — APPROVED** [review]
Very cool, see my suggestion, I think we should use `object`, it's broader and applicable to more cases and seems to work just fine

File: src/_types.ts (hunk 1 above)
Comment (dartess) [reply src/_types.ts]: It doesn't matter at the moment, since the T sent always matches `extends Record<PropertyKey, unknown>`. But it might be useful in the future.

**Comment by TkDodo** [comment]
could you please add a type-test that fails with the current version of the function but passes on this PR? Thanks.

**Comment by dartess** [comment]
@TkDodo I don't see any type tests in this package. do you want me to take on the task of setting it up?

**Comment by eranhirsch** [comment]
@dartess, here are example of typing tests we have already. We use vitest's `expectTypeOf`:

https://github.com/search?q=repo%3Aremeda%2Fremeda%20expectTypeOf&type=code

**Comment by dartess** [comment]
@eranhirsch cool, I didn't know vitest could do that. I will add tests soon.

**Comment by dartess** [comment]
@TkDodo done!

File: src/mapValues.test.ts [hunk 2]
```diff
@@ -34,3 +34,27 @@ describe('data last', () => {
...
+
+  test('indexed signature', () => {
+    mapValues(
+      // eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style -- must be "indexed signature"
+      {} as { [key: string]: unknown },
+      (_, key) => {
+        expectTypeOf(key).toEqualTypeOf<string>();
+      }
+    );
+  });
```
Comment (eranhirsch) [inline src/mapValues.test.ts:59]: Just to make sure we cover all our bases, can you add tests for:
* `{ [key in number]: unknown }`
* `{ [key: number]: unknown }`
* `{ [key in `prefix${string}`]: unknown }`
* `{ [key: `prefix${string}`]: unknown }`

**Review by eranhirsch — APPROVED** [review]
Looks great

File: src/mapValues.test.ts (hunk 2 above)
Comment (dartess) [reply src/mapValues.test.ts]: good, also add tests for symbols

File: src/mapValues.test.ts [hunk 3]
```diff
@@ -34,3 +34,51 @@ describe('data last', () => {
...
+    mapValues({} as { [K in number]: unknown }, (_, key) => {
+      expectTypeOf(key).toEqualTypeOf<`${number}`>();
+    });
+    mapValues({} as { [K in `prefix${string}`]: unknown }, (_, key) => {
+      expectTypeOf(key).toEqualTypeOf<`prefix${string}`>();
+    });
+    mapValues({} as { [K in symbol]: unknown }, (_, key) => {
+      expectTypeOf(key).toEqualTypeOf<never>();
+    });
+  });
```
Comment (eranhirsch) [inline src/mapValues.test.ts:66]: nit: I prefer using a `describe` here and breaking each assert to a separate test so that we get fine-tuned test results with the specific issue. Up to you if you want to change it.

**Review by eranhirsch — APPROVED** [review]
Perfection

File: src/mapValues.test.ts [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import { mapValues } from './mapValues';
 import { pipe } from './pipe';
+import { describe } from 'vitest';
```
Comment (eranhirsch) [inline src/mapValues.test.ts:3]: Hmmm... Why is this needed, we have global types for vitest, the file already used describe without it
